crates/rpc/rpc-builder/src/lib.rs

```diff
@@ -1634,139 +1634,48 @@ impl RpcServerConfig {
-
-        let ws_socket_addr = self
-            .ws_addr
-            .unwrap_or(SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, DEFAULT_WS_RPC_PORT)));
-
-        // If both are configured on the same port, we combine them into one server.
-        if self.http_addr == self.ws_addr &&
-            self.http_server_config.is_some() &&
-            self.ws_server_config.is_some()
-        {
-            let cors = match (self.ws_cors_domains.as_ref(), self.http_cors_domains.as_ref()) {
-                (Some(ws_cors), Some(http_cors)) => {
-                    if ws_cors.trim() != http_cors.trim() {
-                        return Err(WsHttpSamePortError::ConflictingCorsDomains {
-                            http_cors_domains: Some(http_cors.clone()),
-                            ws_cors_domains: Some(ws_cors.clone()),
-                        }
-                        .into())
-                    }
-                    Some(ws_cors)
-                }
-                (a, b) => a.or(b),
-            }
-            .cloned();
-
-            let secret = self.jwt_secret.clone();
-
-            // we merge this into one server using the http setup
-            self.ws_server_config.take();
-
-            modules.config.ensure_ws_http_identical()?;
-
-            let builder = self.http_server_config.take().expect("http_server_config is Some");
-            let (server, addr) = WsHttpServerKind::build(
-                builder,
-                http_socket_addr,
-                cors,
-                secret,
-                ServerKind::WsHttp(http_socket_addr),
-                modules
-                    .http
-                    .as_ref()
-                    .or(modules.ws.as_ref())
-                    .map(RpcRequestMetrics::same_port)
-                    .unwrap_or_default(),
-            )
-            .await?;
-            return Ok(WsHttpServer {
-                http_local_addr: Some(addr),
-                ws_local_addr: Some(addr),
-                server: WsHttpServers::SamePort(server),
-                jwt_secret,
-            })
-        }
-
-        let mut http_local_addr = None;
-        let mut http_server = None;
-
-        let mut ws_local_addr = None;
-        let mut ws_server = None;
-        if let Some(builder) = self.ws_server_config.take() {
-            let builder = builder.ws_only();
-            let (server, addr) = WsHttpServerKind::build(
-                builder,
-                ws_socket_addr,
-                self.ws_cors_domains.take(),
-                self.jwt_secret.clone(),
-                ServerKind::WS(ws_socket_addr),
-                modules.ws.as_ref().map(RpcRequestMetrics::ws).unwrap_or_default(),
-            )
-            .await?;
-            ws_local_addr = Some(addr);
-            ws_server = Some(server);
+    /// Sets up and starts a server with optional CORS, JWT authentication, and metrics at `addr`.
+    async fn launch_server(
+        addr: SocketAddr,
```

Comment:
this needs to take `self` and use all the configured fields

and then we also need the previous logic that checks

> If both are configured on the same port, we combine them into one server.

optional layer can be configured via an optional layer

https://docs.rs/tower/latest/tower/builder/struct.ServiceBuilder.html#method.option_layer

the motivation behind this pr is to compress the previous two steps (with the intermediary enum) into one function